File: test/language/templates/java/templates/StructRecursiveTemplateTest.java
package templates;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.Test;

import java.io.IOException;

import zserio.runtime.io.BitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamWriter;

import templates.struct_recursive_template.StructRecursiveTemplate;
import templates.struct_recursive_template.Compound_uint32;
import templates.struct_recursive_template.Compound_Compound_uint32;
import templates.struct_recursive_template.Compound_string;
import templates.struct_recursive_template.Compound_Compound_string;
import templates.struct_recursive_template.Compound_Compound_Compound_string;

public class StructRecursiveTemplateTest
{
    @Test
    public void readWrite() throws IOException
    {
        final StructRecursiveTemplate structRecursiveTemplate = new StructRecursiveTemplate();
        structRecursiveTemplate.setCompound1(new Compound_Compound_uint32(
                new Compound_uint32(42)));
        structRecursiveTemplate.setCompound2(new Compound_Compound_Compound_string(
                new Compound_Compound_string(new Compound_string("string"))));

        final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        structRecursiveTemplate.write(writer);

        final BitStreamReader reader = new ByteArrayBitStreamReader(
                writer.toByteArray(), writer.getBitPosition());
        final StructRecursiveTemplate readStructRecursiveTemplate = new StructRecursiveTemplate(reader);
        assertTrue(structRecursiveTemplate.equals(readStructRecursiveTemplate));
    }
}

File: test/extensions/language/templates/cpp/StructRecursiveTemplateTest.cpp
#include "gtest/gtest.h"
#include "templates/struct_recursive_template/StructRecursiveTemplate.h"
#include "zserio/RebindAlloc.h"

namespace templates
{
namespace struct_recursive_template
{

using allocator_type = StructRecursiveTemplate::allocator_type;
using string_type = zserio::string<allocator_type>;

TEST(StructRecursiveTemplateTest, readWrite)
{
    StructRecursiveTemplate structRecursiveTemplate;
    structRecursiveTemplate.setCompound1(Compound_Compound_uint32{Compound_uint32{42}});
    structRecursiveTemplate.setCompound2(Compound_Compound_Compound_string{
            Compound_Compound_string{Compound_string{string_type{"string"}}}});

    zserio::BitBuffer bitBuffer = zserio::BitBuffer(1024 * 8);
    zserio::BitStreamWriter writer(bitBuffer);
    structRecursiveTemplate.write(writer);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    StructRecursiveTemplate readStructRecursiveTemplate(reader);

    ASSERT_TRUE(structRecursiveTemplate == readStructRecursiveTemplate);
}

} // namespace struct_recursive_template
} // namespace templates
